cmd/snap/cmd_snapshot.go

```diff
@@ -0,0 +1,303 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/strutil/quantity"
+)
+
+type timeOpts struct {
+	AbsTime bool `long:"absolute-times"`
+	RelTime bool `long:"relative-times"`
+}
+
+var timeDescs = mixinDescs{
+	"absolute-times": i18n.G("Always display absolute times (in YYYY-MM-DD HH:MM format)."),
+	"relative-times": i18n.G("Always display relative times. If neither absolute nor relative times are requested, relative times are used for up to 30 days, and then absolute times in YYYY-MM-DD format."),
+}
+
+func fmtSize(size int64) string {
+	return quantity.FormatAmount(uint64(size), -1)
+}
+
+func (opt timeOpts) fmtTime(t time.Time) string {
+	if opt.AbsTime && !opt.RelTime {
+		return t.Round(time.Minute).Format("2006-01-02 15:04")
+	}
+	ago := time.Since(t)
+	if (opt.RelTime && !opt.AbsTime) || ago < 30*24*time.Hour {
+		// TRANSLATORS: %s is a (separately translated) compact representation of a duration (e.g. 1h30m)
+		return fmt.Sprintf(i18n.G("%s ago"), quantity.FormatDuration(ago.Seconds()))
+	}
+
+	return t.Round(24 * time.Hour).Format("2006-01-02")
+}
+
+type savedCmd struct {
+	timeOpts
+	Wide       bool       `long:"wide"`
+	ID         snapshotID `long:"id"`
+	Positional struct {
+		Snaps []installedSnapName `positional-arg-name:"<snap>"`
+	} `positional-args:"yes"`
+}
+
+var shortSavedHelp = i18n.G("List currently stored snapshots")
+var longSavedHelp = i18n.G(`
+The saved command lists the snapshots that have been created previously with the 'save' command.
+`)
+
+func (x *savedCmd) tabline(sg *client.SnapshotGroup, extraWidth int) string {
+	if len(sg.Snapshots) == 0 {
+		return fmt.Sprintf("%d\t-\t-\t-", sg.ID)
+	}
+	mint := sg.MinTime()
+	snapNames := make([]string, 0, len(sg.Snapshots))
+	var sz int64
+	for _, sh := range sg.Snapshots {
+		snapNames = append(snapNames, sh.Snap)
+		sz += sh.Size
+	}
+	snaps := strings.Join(snapNames, ", ")
+	if !x.Wide {
+		width, _ := termSize()
+		// size (5) + gutters (3+2+2; why is the first gutter 3?)
+		width -= 12 + extraWidth
+		if len(snaps) > width {
+			snaps = snaps[:width-1] + "…"
```

Comment:
We should not cut a snap name in half here.. we already have constraints on maximum name length, so in the worst case we get "foobar,..." which should be acceptable. The logic should count each name until the total exceeds a given limit, and then cut `snapNames` itself to N entries.

Reply:
good idea, will do.